include/aspect/introspection.h

```diff
@@ -84,12 +107,29 @@ namespace aspect
       const unsigned int n_blocks;
 
       /**
+       * A structure that enumerates the vector blocks of the finite element
+       * that correspond to each of the variables in this problem.
+       */
+      struct BlockIndices
+      {
+        unsigned int       velocities;
+        unsigned int       pressure;
+        unsigned int       temperature;
+        std::vector<unsigned int> compositional_fields;
+      };
+      /**
+       * A variable that enumerates the vector blocks of the finite element
+       * that correspond to each of the variables in this problem.
+       */
+      const BlockIndices block_indices;
+
+      /**
        * A structure that contains FEValues extractors for every block of the
        * finite element used in the overall description.
        */
       struct Extractors
       {
-        Extractors (const unsigned int n_compositional_fields);
+        Extractors (const Introspection<dim>::ComponentIndices &component_indices);
```

Comment:
I think that if you write it like this, C++98 requires a `typename`. But, I also think that the members of the parent are accessible without qualification, so you could just write `ComponentIndices` without the prefix qualification.